[ci-app] Cypress Plugin Instrumentation

juan-fernandez opened the pull request:
### What does this PR do?
This is a prototype for a `cypress` plugin. 

The way the user would instrument `cypress` would be the following:

On `cypress`'s [`pluginsFile`](https://docs.cypress.io/guides/references/configuration#Folders-Files): 
```javascript
// By default cypress/plugins/index.js
module.exports = require('dd-trace/cypress/plugin')
```

Or if they only use our plugin they can set in their [`cypress.json`](https://docs.cypress.io/guides/references/configuration#cypress-json):

```json
{
  "pluginsFile": "dd-trace/cypress/plugin"
}
```


And on `cypress`'s [`supportFile`](https://docs.cypress.io/guides/references/configuration#Folders-Files):
```javascript
// By default cypress/support/index.js
require('dd-trace/cypress/support')
```

### Motivation
Instrument `cypress`

### Plugin Checklist
<!-- Fill this section if adding or updating a plugin. Remove otherwise. -->

- [x] Unit tests.
- [x] TypeScript [definitions][1].
- [x] TypeScript [tests][2].
- [ ] API [documentation][3].
- [x] CircleCI [jobs/workflows][4].
- [x] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Additional Notes
<!-- Anything else we should know when reviewing? -->

rochdev commented on packages/datadog-plugin-cypress/test/index.spec.js:
```diff
@@ -0,0 +1,73 @@
+  let agentListenPort
+  withVersions(plugin, ['cypress'], (version, moduleName) => {
+    beforeEach(() => {
+      return agent.load(['cypress']).then((agentPort) => {
```
Is this even needed if there is no automatic instrumentation?

juan-fernandez replied:
we need to spin up a "fake" agent to read traces from. We probably don't need to load any specific instrumentation though

rochdev commented on packages/datadog-plugin-cypress/test/index.spec.js:
```diff
@@ -0,0 +1,73 @@
+      })
+    })
+    afterEach(() => {
+      return Promise.all([
```
This would be a bit easier to read as 2 blocks, for example `afterEach(() => agent.close())` and `afterEach(done => appServer.close(done))`.

juan-fernandez replied:
yeah, that's right 👍

rochdev commented on packages/dd-trace/test/plugins/agent.js:
```diff
@@ -41,7 +41,7 @@ module.exports = {
         server.on('connection', socket => sockets.push(socket))
 
-        listener = server.listen(port, 'localhost', resolve)
+        listener = server.listen(port, 'localhost', () => resolve(port))
```
Why is it needed?

juan-fernandez replied:
so we need to be able to tell the tracer running in the cypress plugin process which port to report to. This is not needed when testing within process, but the tracer is now being run in a different process

juan-fernandez commented on packages/datadog-plugin-cypress/src/plugin.js:
```diff
@@ -0,0 +1,88 @@
+  }
+}
+
+module.exports = (on, config) => {
```
what would the code look like and what would the user have to do?

juan-fernandez replied:
so as mentioned: let's abandon this idea for the moment since it is risky and probably requires a big change

juan-fernandez commented on packages/datadog-plugin-cypress/src/plugin.js:
```diff
@@ -0,0 +1,88 @@
+      } = getTestSpanMetadata(tracer, testName, testSuite)
+
+      if (!activeSpan) {
+        activeSpan = tracer.startSpan('cypress.test', {
```
there won't be. The interesting stuff is happening at the browser, which is a different process.

juan-fernandez replied:
actually, in some small subset of cases the user might be doing something worth instrumenting _within_ the plugin process. I'd say that's unlikely though

juan-fernandez replied:
thinking about this more, I know we talked about activating the scope, but how would I do it? This is a `beforeEach` hook, so it's not like I can wrap the test function or anything. Am I missing something? 🤔

juan-fernandez commented on packages/datadog-plugin-cypress/src/plugin.js:
```diff
@@ -0,0 +1,88 @@
+      const { state, error } = test
+      if (activeSpan) {
+        activeSpan.setTag(TEST_STATUS, CYPRESS_STATUS_TO_TEST_STATUS[state])
+        if (error) {
```
oh of course, I don't know why I put it this way. Thanks!

juan-fernandez commented on packages/datadog-plugin-cypress/src/support.js:
```diff
@@ -0,0 +1,17 @@
+/* eslint-disable */
+beforeEach(() => {
+  cy.task('beforeEach', {
```
good call. Yeah, we created it just for our instrumentation. I'll add a prefix 👍

juan-fernandez commented on packages/datadog-plugin-cypress/test/app/cypress/plugins/index.js:
```diff
@@ -0,0 +1,7 @@
+module.exports = (on, config) => {
+  // We can't use the tracer available in the testing process, because this code is
+  // run in a different process. We need to init a different tracer reporting to the
+  // url set by the plugin agent
```
it is not standard. This plugin code runs in a different process (`spawn`) and does not inherit (AFAICT) env vars from the main process. If you want variables available here you have to go through `config.env`

juan-fernandez replied:
> Also, why force disable startup logs and all plugins? It's possible some of them would work, especially if this code runs before the test files are loaded.

Yeah this might be interesting if users are using libraries we instrument within the plugin process.

juan-fernandez replied:
you're absolutely right about using the existing environment variables 👍 . As for the plugins: it might be interesting if the user is doing stuff _within_ the plugin process with libraries we instrument.

juan-fernandez commented on packages/datadog-plugin-cypress/src/plugin.js:
```diff
@@ -0,0 +1,88 @@
+}
+
+function getTestSpanMetadata (tracer, testName, testSuite) {
+  const childOf = tracer.extract('text_map', {
```
how would you call this helper function? `getTestParentSpan` ? That seems too specific, but we only use it for tests, right?

juan-fernandez commented on packages/dd-trace/test/plugins/agent.js:
```diff
@@ -41,7 +41,7 @@ module.exports = {
         server.on('connection', socket => sockets.push(socket))
 
-        listener = server.listen(port, 'localhost', resolve)
+        listener = server.listen(port, 'localhost', () => resolve(server))
```
I can't do the `agent.server = server` we mentioned because what we require is this file, not `agent`. I'd have to do add a `getAgent()` function to the `module.exports` in this file, which seems worse. 
We also call `agent` to the result of the `require` of this file, so we'd have to rename that in every file. 

I think this accomplishes the flexibility

juan-fernandez commented on cypress/plugin.js:
```diff
@@ -0,0 +1,5 @@
+require('..').init({
```
these files are now easier to export, but I don't know if this scales well: a folder per plugin?

rochdev replied:
I think it's fine since we won't need to do this often, and eventually plugins will be exported in a different way.

juan-fernandez commented:
so I've addressed the feedback from yesterday @rochdev. These pieces would be missing: 

* Activate the scope in the cypress plugin. I don't know if we can do that with the hook based instrumentation we have. 
* The new path of the cypress plugin might not be the best: `cypress/plugin.js`. It would mean that we would grow the home folder for each plugin, if we want to move to this approach. Ideas?

juan-fernandez commented:
@rochdev do the latest changes look OK?

rochdev reviewed (changes requested):
A few last small things and then good to go 👍

juan-fernandez commented on packages/datadog-plugin-cypress/src/index.js:
```diff
@@ -0,0 +1,6 @@
+module.exports = [
```
yeah I'll do that

rochdev replied:
Ah yes I was thinking that externals.json could be used for this but it needs at least the version of the actual library to get started. I guess it's fine then, it will simply never be loaded.

juan-fernandez replied:
I moved it to the test file in https://github.com/DataDog/dd-trace-js/commit/a1d54686ca5ab26096aac2396220051a354e2a07, but if you think it works there I can revert it

rochdev replied:
Fine either way so let's keep it in the test 👍

juan-fernandez commented:
@rochdev any other concern? I think the tests failing are also failing in master.

juan-fernandez commented:
there are tests failing in windows. I'm trying to figure out why

juan-fernandez commented:
A last set of changes: 
* I went back to the old version of `get-port` because the new one seems to fail for windows. It didn't add much (other than reduce the risk of port collision which is low anyway): https://github.com/DataDog/dd-trace-js/commit/8189e353dda7bf21f81d04e058fc3b3003ffece9 
* I had a typo where I was using `cucumber` plugin for the `cypress` one: https://github.com/DataDog/dd-trace-js/commit/de216c37a07b206483e52ad8ae308abe8d8c2b7c#diff-3b357b1cc0d572d24888287611233e658c09559c3633e5396459c2f58ace9cbfL13 
* This made me realise that it's safer to have the `datadog-plugin-cypress/src/index.js` because we assume a `index.js` is there at least in `packages/dd-trace/src/plugins/index.js`, so I added it back.

juan-fernandez commented:
it doesn't seem like the `node-aws-sdk` tests failing is related to my changes. Maybe they're failing because of the latest release 17 hours ago? https://github.com/aws/aws-sdk-js/releases/tag/v2.990.0

rochdev commented:
> it doesn't seem like the node-aws-sdk tests failing is related to my changes. Maybe they're failing because of the latest release 17 hours ago?

It does look like it.
